main/guides/js-programming/hardened-js.md

```diff
@@ -0,0 +1,388 @@
+
+[![makeCounter code animation](../../assets/counter-animation.png)](https://pythontutor.com/live.html#code=const%20makeCounter%20%3D%20%28%29%20%3D%3E%20%7B%0A%20%20let%20count%20%3D%200%3B%0A%20%20const%20it%20%3D%20%7B%0A%20%20%20%20incr%3A%20%28%29%20%3D%3E%20%7B%0A%20%20%20%20%20%20count%20%2B%3D%201%3B%0A%20%20%20%20%20%20return%20count%3B%0A%20%20%20%20%7D,%0A%20%20%20%20decr%3A%20%28%29%20%3D%3E%20%7B%0A%20%20%20%20%20%20count%20-%3D%201%3B%0A%20%20%20%20%20%20return%20count%3B%0A%20%20%20%20%7D,%0A%20%20%7D%3B%0A%20%20return%20Object.freeze%28it%29%3B%0A%7D%3B%0A%0Aconst%20c1%20%3D%20makeCounter%28%29%3B%0A%0Aconst%20c2%20%3D%20makeCounter%28%29%3B%0Aconsole.log%28c2.incr%28%29%29%3B%0Aconsole.log%28%5Bc1.incr%28%29,%20c2.incr%28%29%5D%29%3B&cumulative=false&curInstr=25&heapPrimitives=nevernest&mode=display&origin=opt-live.js&py=js&rawInputLstJSON=%5B%5D&textReferences=false")
+
+## Hardening JavaScript: strict mode
+
+The first step to hardening JavaScript is that Hardened JavaScript
+is always in [strict mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode).
+
+[![Subsetting JavaScript](https://raw.githubusercontent.com/endojs/Jessie/main/docs/jessie.png)](https://github.com/endojs/Jessie#subsetting-ecmascript)
+
+The most likely way that you would notice this is if you
+make a typo in a variable name: this will throw a `ReferenceError`
+rather than silently continuing, with untold results.
+
+Important benefits of strict mode include complete encapsulation
+(no `caller` etc.) and reliable static scoping.
+
+## Hardening JavaScript: frozen built-ins
+
+One form of authority that is too widely available in
+ordinary JavaScript is the ability to redefine built-ins
+(shown above as "mutable primordials").
+Consider this `changePassword` function:
+
+<<< @/snippets/test-no-ses.js#changePassword
+
+In ordinary JavaScript, since someone might have redefined
+the `includes` method on `Array` objects, we run the risk of stolen passwords:
+
+<<< @/snippets/test-no-ses.js#exfiltrate
+
+In Hardened JavaScript, the `Object.assign` fails because `Array.prototype` and all other
+[standard, built-in objects](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects)
+are immutable.
+
+::: tip Compatibility issues with `ses` / Hardened JavaScript
+Certain libraries that make tweaks to the standard built-ins
+may fail in Hardened JavaScript.
+
+The [SES wiki](https://github.com/endojs/endo/wiki) tracks compatibility
+reports for NPM packages, including potential workarounds.
```

Comment:
It does indeed. I didn't know about this and have a separate tracking bug for this purpose. Outside of this PR, we should consolidate.

Reply:
The tracking bug is https://github.com/endojs/endo/issues/576
I made a note there: https://github.com/endojs/endo/issues/576#issuecomment-986064344